language/testing-infra/functional-tests/src/testsuite.rs

```diff
@@ -65,31 +80,87 @@ fn print_stage(haystack: &str) -> bool {
         .unwrap_or(true)
 }
 
-// Runs all tests under the test/testsuite directory.
-pub fn functional_tests<TComp: Compiler>(
-    compiler: TComp,
-    path: &Path,
+fn write_horizontal_line(output: &mut Buffer, term_width: usize) -> std::io::Result<()> {
+    writeln!(
+        output,
+        "{}",
+        iter::repeat('=').take(term_width).collect::<String>()
+    )
+}
+
+fn write_test_header(
+    output: &mut Buffer,
+    term_width: usize,
+    test_file_path: &Path,
+) -> std::io::Result<()> {
+    writeln!(output)?;
+    write_horizontal_line(output, term_width)?;
+    writeln!(output, "{}", test_file_path.display())?;
+    writeln!(output)
+}
+
+fn check_or_update_expected_output(
+    bufwtr: &BufferWriter,
+    output: &mut Buffer,
+    term_width: usize,
+    test_file_path: &Path,
+    exp_file_path: &Path,
+    log: &EvaluationLog,
 ) -> datatest_stable::Result<()> {
-    let input = read_to_string(path)?;
+    // REVIEW: "\n" is probably not very portable.
```

Comment:
why not?

Reply:
Well the usual "\r" vs "\r\n" vs "\n" thing. Mac uses "\r" but understands "\n". Windows uses "\r\n" and sometimes only understands this. Linux sometimes only understands "\n". Just saying there's a potential risk of the result of the comparison being platform dependent and we should be aware of it. Ideally I should let the library handle this... Hmm maybe I can? Let me see if I can use a string buffer and then `wrtieln!` to it.

Reply:
I don't think this really matters. We only have `\n` throughout our rust codebase, I think its fine to carry that over to move and exp files